lgc/patch/LowerDebugPrintf.cpp

```diff
@@ -0,0 +1,239 @@
+/*
+************************************************************************************************************************
+*
+*  Copyright (C) 2017-2021 Advanced Micro Devices, Inc. All rights reserved.
+*
+***********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  LowerDebugPrintf.cpp
+ * @brief LLPC source file: contains implementation of class lgc::LowerDebugPrintf.
+ ***********************************************************************************************************************
+ */
+#include "lgc/patch/LowerDebugPrintf.h"
+#include "lgc/patch/Patch.h"
+#include "lgc/state/PalMetadata.h"
+#include "lgc/state/PipelineState.h"
+#include "llvm/ADT/Hashing.h"
+#include "llvm/BinaryFormat/MsgPackDocument.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "lgc-patch-image-op-collect"
+
+using namespace llvm;
+using namespace lgc;
+
+namespace lgc {
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+// @returns : The preserved analyses (The analyses that are still valid after this pass)
+PreservedAnalyses LowerDebugPrintf::run(Module &module, ModuleAnalysisManager &analysisManager) {
+  LLVM_DEBUG(dbgs() << "Run the pass Lower-debug-printf\n");
+  PipelineState *pipelineState = analysisManager.getResult<PipelineStateWrapper>(module).getPipelineState();
+  Patch::init(&module);
+  m_pipelineState = pipelineState;
+  SmallVector<CallInst *, 16> callees;
+  BuilderBase builder(*m_context);
+  for (auto &func : module) {
+    auto name = func.getName();
+    if (name.startswith(lgcName::LowerDebugPrintf)) {
+      for (auto user : func.users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          builder.SetInsertPoint(callInst);
+          auto resultVal = createDebugPrintf(callInst->getArgOperand(0), callInst->getArgOperand(1),
+                                             make_range(callInst->arg_begin() + 2, callInst->arg_end()), builder);
+          callInst->replaceAllUsesWith(resultVal);
+          callees.push_back(callInst);
+        }
+      }
+    }
+  }
+
+  for (auto callInst : callees) {
+    callInst->dropAllReferences();
```

Comment:
I would hope that the `dropAllReferences` is not needed here, because all uses have been replaced above.